Make LinearCoord working

fernandezc opened the pull request:
Many functions were not working with LinearCoord. This is a PR trying to solve these problems

fernandezc commented:
@atravert I sometimes have an error with IRIS test due to a G matrix not positively definite. I tried to solve the problem in on of the commit above https://github.com/spectrochempy/spectrochempy/pull/245/commits/ac52f9b664755188f19530f82f644d42ae1dc151
Can you tell me if this is ok or should not be done?  because I did not enter the details of this algorithm.

fernandezc commented:
One error happen in "test_basecor_sequential" only on Travis and python 3.9 configuration, not on osx.

atravert commented:
> 
> 
> @atravert I sometimes have an error with IRIS test due to a G matrix not positively definite. I tried to solve the problem in the commit above [ac52f9b](https://github.com/spectrochempy/spectrochempy/commit/ac52f9b664755188f19530f82f644d42ae1dc151)
> Can you tell me if this is ok or should not be done? because I did not enter the details of this algorithm.

strange; this seems to be an old version of  iris.py. Some tilmes ago I have added nearestPD() a function ensuring that G is positive definite. This part of the code now looks like this: 
```
                fi = np.zeros((len(eps), len(X.x.data)))
                if verbose:
                    print('... Solving for lambda = {} ...'.format(lamda))
                G = nearestPD(G0 + 2 * lamda * S)
                for j, freq in enumerate(X.x.data):
                    fi[:, j] = quadprog.solve_qp(G, a[j].squeeze(), C, b)[0]
```

fernandezc commented:
> > @atravert I sometimes have an error with IRIS test due to a G matrix not positively definite. I tried to solve the problem in the commit above [ac52f9b](https://github.com/spectrochempy/spectrochempy/commit/ac52f9b664755188f19530f82f644d42ae1dc151)
> > Can you tell me if this is ok or should not be done? because I did not enter the details of this algorithm.
> 
> strange; this seems to be an old version of iris.py. Some tilmes ago I have added nearestPD() a function ensuring that G is positive definite. This part of the code now looks like this:
> 
> ```
>                 fi = np.zeros((len(eps), len(X.x.data)))
>                 if verbose:
>                     print('... Solving for lambda = {} ...'.format(lamda))
>                 G = nearestPD(G0 + 2 * lamda * S)
>                 for j, freq in enumerate(X.x.data):
>                     fi[:, j] = quadprog.solve_qp(G, a[j].squeeze(), C, b)[0]
> ```

Yes, this is what I had... I just added one line
```python

                fi = np.zeros((len(eps), len(X.x.data)))
                if verbose:
                    print('... Solving for lambda = {} ...'.format(lamda))
                G = nearestPD(G0 + 2 * lamda * S)

                # The following line is to avoid ValueError: 'matrix G is not
                # positive definite'
                # SEE: https://github.com/facebookresearch/GradientEpisodicMemory/issues/2#issuecomment-431826393
                G += G * 0.00001

                for j, freq in enumerate(X.x.data):
                    fi[:, j] = quadprog.solve_qp(G, a[j].squeeze(), C, b)[0]

```

This problem actually is erratic, however... sometimes it works sometimes not.

atravert commented:
Ah OK, sorry. So then the problem is probably with `isPD()` (on the 
bottom of iris.py)  which probably does not test positive-definiveness 
as in quadprog() (where it is defined as an error returbed by the c 
fucntion which actually makes the qp optimization: line 
https://github.com/quadprog/quadprog/blob/b166c2ba6299c78001634a6a1745bb7213af86a5/quadprog/quadprog.pyx)

If the fix works, let's |go for it !| ||

Le 21/02/2021 à 09:30, Christian Fernandez a écrit :
>
>         @atravert <https://github.com/atravert> I sometimes have an
>         error with IRIS test due to a G matrix not positively
>         definite. I tried to solve the problem in the commit above
>         ac52f9b
>         <https://github.com/spectrochempy/spectrochempy/commit/ac52f9b664755188f19530f82f644d42ae1dc151>
>         Can you tell me if this is ok or should not be done? because I
>         did not enter the details of this algorithm.
>
>     strange; this seems to be an old version of iris.py. Some tilmes
>     ago I have added nearestPD() a function ensuring that G is
>     positive definite. This part of the code now looks like this:
>
>     |fi = np.zeros((len(eps), len(X.x.data))) if verbose: print('...
>     Solving for lambda = {} ...'.format(lamda)) G = nearestPD(G0 + 2 *
>     lamda * S) for j, freq in enumerate(X.x.data): fi[:, j] =
>     quadprog.solve_qp(G, a[j].squeeze(), C, b)[0] |
>
> Yes, this is what I had... I just added one line
>
>                  fi  =  np.zeros((len(eps),len(X.x.data)))
>                  if  verbose:
>                      print('... Solving for lambda = {} ...'.format(lamda))
>                  G  =  nearestPD(G0  +  2  *  lamda  *  S)
>
>                  # The following line is to avoid ValueError: 'matrix G is not
>                  # positive definite'
>                  # SEE: 
> https://github.com/facebookresearch/GradientEpisodicMemory/issues/2#issuecomment-431826393
>                  G  +=  G  *  0.00001
>
>                  for  j,freq  in  enumerate(X.x.data):
>                      fi[:,j]=  quadprog.solve_qp(G,a[j].squeeze(),C,b)[0]
>
> This problem actually is erratic, however... sometimes it works 
> sometimes not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/spectrochempy/spectrochempy/pull/245#issuecomment-782819935>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ACNBNCW65SPE745EZBPCVQ3TADAB3ANCNFSM4X5ZK27A>.
>
